### PR TITLE
Fix fill_buf by not calling poll_fill_buf twice

### PR DESCRIPTION
**Pull request by Darksonn**
There are some IO resources which hit the panic when calling `poll_fill_buf` twice, e.g. a `BufReader<File>` will do this when you are at EOF.

A safe alternative is to check for empty slices and handle those in a special way, but I prefer this as it is a more robust solution.

Fixes: #4085

### CODE REVIEW

**Comment by UebelAndre** [comment]
Should some kind of regression test be added here?

**Comment by Darksonn** [comment]
I don't think it's particularly necessary here, but I can add one.

**Review by carllerche — APPROVED** [review]

File: tokio/tests/io_fill_buf.rs [hunk 1]
```diff
@@ -0,0 +1,34 @@
+#![warn(rust_2018_idioms)]
+#![cfg(feature = "full")]
+
+use tempfile::NamedTempFile;
+use tokio::fs::File;
+use tokio::io::{AsyncBufReadExt, BufReader};
+use tokio_test::assert_ok;
+
+#[tokio::test]
+async fn fill_buf_file() {
```
Comment (hawkw) [inline tokio/tests/io_fill_buf.rs:10]: take it or leave it: maybe worth a comment summarizing the bug this is a regression test for (and a link to the original issue?)

**Review by hawkw — APPROVED** [review]

File: tokio/tests/io_fill_buf.rs (hunk 1 above)
Comment (Darksonn) [reply tokio/tests/io_fill_buf.rs:10]: I don't really feel like waiting for CI over again, since I am making a release immediately after merging this.

**Pull request merged**